Add support for CA/NY/LA/VA/HI/UT/NJ/MN/DE/CT/NV/NM/OK/KY/GU/MP/YT/BC/SK/AB/NT/NS/MB/ON/NL/PE/Cayman Islands vaccine records, other changes

steven676 opened the pull request:
This patch series makes a number of changes:

* Make signature verification optional in the webapp, which makes debugging and casual inspection easier.
* Add support for detecting which keys to use to verify SHC signatures using the iss field.
* Enable signature verification for California digital vaccine records (closes #4)
* Various quality-of-life improvements: stop the camera in the webapp once a barcode is read; behave better when QR codes that aren't valid SHCs are presented.

Note that this patch series currently (probably) breaks signature verification of Quebec vaccination proofs -- I need to fill in the iss value for Quebec at https://github.com/steven676/shc-covid19-decoder/blob/main/src/issuers.js#L4, which I can't find anywhere public.

Lightly tested with valid California vaccine records, the sample record at https://github.com/dvci/health-cards-walkthrough/blob/main/SMART%20Health%20Cards.ipynb, and a number of non-SHC QR codes.

steven676 committed:
Display only verified data when signature verification is enabled

steven676 committed:
Webapp: indicate in the output whether issuer signature was verified

steven676 committed:
Webapp: make signature verification optional

steven676 committed:
Move issuer public key into a separate module

This is in preparation for the ability to handle multiple issuers.

steven676 committed:
decodeJWS(): return the entire decoded blob, not just the vaccination record

We'll need this to detect the issuer.

steven676 committed:
Use the issuer field of the SHC to select signing keys for verification

This will allow us to verify SHCs from multiple issuers.  Note that this
requires decoding the data before verifying it; this is probably a bad
idea (similar to violating Moxie Marlinspike's rule of thumb that MAC
verification should come before any other receive-side operations in a
protocol [1]), but well, we're stuck with this format ...

[1] https://moxie.org/2011/12/13/the-cryptographic-doom-principle.html

steven676 committed:
Add signing key for California Digital COVID-19 Vaccine Record

Key from https://myvaccinerecord.cdph.ca.gov/creds/.well-known/jwks.json:

$ sha256sum jwks.json
eb5d4bbea0734c205daf49c03e3013071c9edfd6232d24451fdfc27c6637df5f  jwks.json

steven676 committed:
Webapp: make debug logging slightly less repetitive

steven676 committed:
dump_shc.js: fix typo

steven676 committed:
More robust error handling

In particular, try to handle cases where corrupt/invalid SMART health
cards, or things that aren't SHCs at all, are scanned.

steven676 committed:
Webapp: stop camera once a barcode is read

thardie commented on src/issuers.js:
```diff
@@ -0,0 +1,26 @@
+    ]
+  },
+  {
+    id: "us.ca",
```
Can you instead use the iss in the provided QR code and download from there? This will make it work for more than just these 2 states then.

steven676 replied:
I thought about this -- it'd certainly be more convenient for debugging and curiosity-driven inspection.  But anyone can generate codes according to the spec, so you'd end up accepting things generated by, say, fakes.antivaxxers.dumb ... For actual verification purposes, you need a list of trusted issuers even if you fetch the keys on the fly from the spec-mandated location.

thardie replied:
Yeah, that totally makes sense. Is there anywhere else collecting a list of trusted issuers? This seems like the start of the common root CAs problem with SSL first came along...

steven676 replied:
There's something calling itself the [Common Trust Network](https://www.commontrustnetwork.org/), which unfortunately seems extremely light on details and doesn't seem to have a documented way of fetching the issuers list.  (If you pull apart [their verifier app](https://thecommonsproject.org/smart-health-card-verifier), you can find an undocumented API endpoint to get the list, but (1) I don't know if they actually want other people using it and (2) it appears to cover US issuers only.)

superay123 replied:
There is also [https://github.com/the-commons-project/vci-directory](url) which has this list [https://raw.githubusercontent.com/the-commons-project/vci-directory/main/vci-issuers.json](url)

steven676 replied:
Oh, that's neat, thanks.  (Still seems to cover US issuers only, though.)

steven676 committed:
dump_shc.js: ensure full contents of SMART health card are displayed

Node.js "helpfully" tries to reduce the amount of output when we pass
the decoded object to console.log(), so we need to pretty-print it
ourselves before displaying.  Fixes #2.  (Thanks to Dave Hughes for the
suggstion to use console.dir() to preserve the color-coding.)

steven676 commented:
A number of changes to my pull request:

* Use the key ID in the header instead of the issuer to select the signing key to verify with.  This avoids having to decode and decompress the payload before verifying it (a bad idea for reasons similar to [the rule of thumb that MAC verification should come before any other receive-side operations in a protocol](https://moxie.org/2011/12/13/the-cryptographic-doom-principle.html)).  As a bonus, demo.png in this repository appears to contain the kid for Quebec's signing key, so verifying Quebec vaccination proofs should work (though I can't test).
* Fix #2 by having dump_shc.js pretty-print the JSON itself instead of leaving it to Node.js's console.log().

steven676 commented:
Thanks for this.  I suppose we should know better than to assume everyone has implemented all of the spec correctly ...  Are your current QR codes signed by the same key as we have in the codebase now, or do the different kid values correspond to different signing keys?

steven676 committed:
Add signing key for New York State Excelsior Pass Plus

Key from https://ekeys.ny.gov/epass/doh/dvc/2021/.well-known/jwks.json:

$ sha256sum jwks.json
b1e12070123acae094fd0047ae4d6b2d8aabe2f1c592932a224e83b83c9d6e2e  jwks.json

steven676 committed:
Add signing key for state of Louisiana LA Wallet

Key from https://healthcardcert.lawallet.com/.well-known/jwks.json:

$ sha256sum jwks.json
fede561fb56d82e82df8a6e5f5f2679460b7a4051a927ec318decf686199b633  jwks.json

steven676 commented:
I added two more signing keys, for the LA Wallet system used by the state of Louisiana and the Excelsior Pass Plus (NOT the Excelsior Pass!) used by the state of New York.

@superay123 can you confirm that it's only the kid that's been changing for the Quebec signing key, not the actual key parameters x/y?  If x/y are stable, I'll pick 4d47866c3941f57f31e3d9b36115d74e641f8847 from your branch into my branch.

superay123 commented:
x/y did not change.

I used the same key (ignoring the changing kid) to verify all our QR 
codes that I downloaded.

I downloaded my QR code again today from the gov site and I verified it 
successfully with the signing key.  The kid and x/y values did not change.

The URL https://covid19.quebec.ca/PreuveVaccinaleApi/issuer/.well-known/jwks.json 
still returns a 404.

If they change the kid or x/y values, i.e. change the signing key, they 
will have to ask everybody to download their QR code again.

I was a bit worried earlier today when @remi and @babelouest said that their kids 
kept changing.

When they downloaded their QR code again, the kid was the one that I 
checked in  4d47866

steven676 commented:
I cherry-picked the kid and iss URL change from @superay123 (with a whitespace adjustment).  Given the instability in Quebec's kid, though, I wonder whether we should go back to my original approach of using the iss URL to identify the signing key (I kept that code around in my [verify-by-issuer branch](https://github.com/steven676/shc-covid19-decoder/tree/verify-by-issuer)).

@superay123 Thanks for the reminder -- I also updated the build at https://steven676.github.io/shc-covid19-decoder/ if anyone wants to test it out.

superay123 commented:
I scanned our QR codes that were downloaded today with [ https://steven676.github.io/shc-covid19-decoder/](url).
Both worked OK

steven676 committed:
Add signing key for Yukon proof of Covid-19 vaccination

Key from https://pvc.service.yukon.ca/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
22e4ad57812178ed419aab7fcbd2c39d083f043b8e9f109e0654436584647ce8  jwks.json

(Found through https://github.com/the-commons-project/vci-directory/pull/60.)

dlh3 commented on dump_shc.js:
```diff
@@ -27,15 +28,23 @@ console.log("-----");
-    return decodeJWS(scannedJWS).then((decoded) => console.log(decoded));
+    return decodeJWSPayload(result.payload).then(
+      (decoded) => {
+        console.log(JSON.stringify(decoded.vc.credentialSubject.fhirBundle.entry, null, 2));
```
```suggestion
        console.dir(decoded.vc.credentialSubject.fhirBundle.entry, {depth:null, compact:false});
```
This fixes #2 (the same problem) but also retains the coloured output.

steven676 replied:
Thanks, that's indeed better -- I'll take care of that in my next round of revisions.

steven676 replied:
Incorporated into a3f626028d6a507219ac11834f093fe12f9521db, thanks!

steven676 committed:
Add signing key for BC Vaccine Card

Key from https://smarthealthcard.phsa.ca/v1/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
dc6c1e25fecbaa4362cd17ede14e862071bc5a0bd74085b7f37a2d369c61e810  jwks.json

With thanks to Mitch Brown and Dave Hughes; closes #14.

steven676 committed:
Update Quebec vaccination proof iss and kid

Thanks to Raymond Ménard for these values.

steven676 commented:
More changes (my apologies for another force push -- this should be the last one):

* Revert to my original approach of selecting the signing key to verify against using the iss URL.  Despite the theoretical advantages of the kid-based approach, the discussion in #12 makes it clear that verifying by iss is likely the only approach that will work with every proof Quebec has issued to date.  Verifying by iss also appears to be what the Commons Project's verifier does.  (For reference, the previous code lives on in my [verify-by-kid branch](https://github.com/steven676/shc-covid19-decoder/tree/verify-by-kid).)
* Add signing keys for Yukon (via https://github.com/the-commons-project/vci-directory/pull/60) and BC (with thanks to @mitchplze and @dlh3; closes #14).

I've also updated my build at https://steven676.github.io/shc-covid19-decoder/ if anyone would like to test it out.

laurencebgood commented:
@steven676 do you have the endpoint for the common trust verifier app? I suppose I could pull it apart myself but you've done the work already so hoping to save some time.

steven676 committed:
Add signing key for Virginia QR code vaccination records

Key from https://apps.vdh.virginia.gov/credentials/.well-known/jwks.json:

$ sha256sum jwks.json
f59ca9bdf677a1476ccf1fb044af24341f3997f9de2f0a83023f23c40a4f0b34  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/92.)

steven676 committed:
Add signing key for Hawaii SMART Health Card

Key from https://travel.hawaii.gov/.well-known/jwks.json:

$ sha256sum jwks.json
64e684830c7d008439a577597126ec5074a966454afb39c9296fb2a8585f7a83  jwks.json

steven676 committed:
Add signing key for eHealth Saskatchewan QR code vaccination records

Saskatchewan appears to have two separate issuers with different signing
keys:

From https://skphr.prd.telushealthspace.com/.well-known/jwks.json:

$ sha256sum jwks.json
96780162f06da7a43218b5fbbb23d8af904c22b8e61717a93435364035770cb4  jwks.json

From https://commons.ehealthsask.ca/.well-known/jwks.json:

$ sha256sum jwks.json
d3c6694b07a74a7954ee63f26c6e8c5c1f2901c9380d80afc032f6b116a0a96b  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/99.)

steven676 commented:
On Wed, Sep 29, 2021 at 03:08:13PM -0700, Laurence B wrote:
> @steven676 do you have the endpoint for the common trust verifier app?
> I suppose I could pull it apart myself but you've done the work
> already so hoping to save some time.

Email me -- it's not exactly hard to find, but out of respect for them
I'd rather not post it openly.

steven676 commented:
Added signing keys for three more governments -- the states of Virginia and Hawaii and the province of Saskatchewan.

@fproulx Do you have any feedback on this that you'd like me to address before merging?

peacey committed:
Add Alberta public keys

lukasbeckerit committed:
Add signing key for NT

[steven@steven676.net: key from https://www.hss.gov.nt.ca/covax/.well-known/jwks.json:
$ sha256sum jwks.json
cf70c7a7405ebbdc99034bb2304c8b19efbb009e759665f620498a9aa00f5b79  jwks.json]

steven676 committed:
Merge "Add Alberta public keys"

Verified by disassembling Alberta's verifier Android app [1], as Alberta
has yet to publish its key to the location specified in the standard.
Additional thanks to /u/YegThrowawayWasTaken on Reddit.

[1] https://play.google.com/store/apps/details?id=ca.ab.gov.covidrecordsverifier

steven676 committed:
Add signing key for Nova Scotia COVID-19 Proof of Vaccination

Key from https://pvc.novascotia.ca/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
8c78b49846b9f86e66b56c6d3a756c08e4a6bce8ff555f3b6dd7039e49847cc0  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/185.)

steven676 committed:
Add signing key for Docket immunization records for Utah

Utah outsources Covid-19 digital vaccine records to Docket [1].

Key from https://docket.care/ut/.well-known/jwks.json:

$ sha256sum jwks.json
15babf40deb6ecd73c9cdf252324eaef34bd042dfafb89dcbbba52dea303b708  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/191.)

[1] https://immunize.utah.gov/usiis/usiis-parents-individuals/

steven676 committed:
Add signing key for Docket immunization records for New Jersey

New Jersey outsources Covid-19 digital vaccine records to Docket [1].

Key from https://docket.care/nj/.well-known/jwks.json:

$ sha256sum jwks.json
9705c17aa5a4ceac5493f69b4dc04e703955dc7b690e7534d66231d759179063  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/191.)

[1] https://covid19.nj.gov/faqs/nj-information/slowing-the-spread/consumer-access-to-covid-19-immunization-records-with-docket-faqs

steven676 committed:
Add signing key for Manitoba Covid-19 immunization card

Key from https://immunizationcard.manitoba.ca/api/national/.well-known/jwks.json:

$ sha256sum jwks.json
0a6388ba092f9aa8460769d170157a4af4c5f80e70166c92a005131522a67335  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/192.)

steven676 committed:
Add signing key for Ontario Covid-19 enhanced vaccination certificate

Key from https://prd.pkey.dhdp.ontariohealth.ca/.well-known/jwks.json:

$ sha256sum jwks.json
a9411fca67636f80260d805b2468980a62f2c309cfcf4e5e9386fdc23a94b2e2  jwks.json

(Found via https://github.com/billylo1/covidpass/commit/bd2ec1a38d5457e8de08adfaac7bc4872ff1c19b;
with thanks to the grassroots vaccine pass team (@grassroots_team on
Twitter)).

steven676 committed:
Add signing key for Cayman Islands Health Services Authority vaccination records

The Cayman Islands appears to use Cerner's electronic medical records
system for all health records, including Covid-19 vaccine records [1].

Key from https://fhir-myrecord.cerner.com/r4/QGFlV8qKdgYu-vPpMAoQW5U4Jb7riiI2/.well-known/jwks.json:

$ sha256sum jwks.json
276845f41ed6cda6c224350649c46fbb22346325f9e53ff183987236bc68dd96  jwks.json

[1] https://www.hsa.ky/our-services/patient-portal-info/

steven676 committed:
Add signing key for Docket immunization records for Minnesota

Minnesota outsources Covid-19 digital vaccine records to Docket [1].

Key from https://docket.care/mn/.well-known/jwks.json:

$ sha256sum jwks.json
d8b09156d15628dfdf411cdba45671b33bd3151eb889aafe0e720919408db44e  jwks.json

[1] https://www.health.state.mn.us/people/immunize/miic/records.html

steven676 commented:
More signing keys:

* Canadian provinces and territories: Alberta (thanks @peacey, closes #16), Northwest Territories (thanks @lpgca), Nova Scotia, Manitoba, Ontario
* US states: Utah, New Jersey, Minnesota (all of which outsource to [Docket](https://docket.care); it's unclear to me whether they're actually issuing SHCs yet, but the NJ/UT submissions to the Commons Project VCI Directory suggest it's coming)
* Cayman Islands Health Services Authority (Cerner user)

steven676 committed:
Add signing key for Newfoundland and Labrador Covid-19 vaccination record

Key from https://www.gov.nl.ca/covid-19/life-during-covid-19/vaccination-record/prod/.well-known/jwks.json:

$ sha256sum jwks.json
66684693005b6f45a1443b53322677e5c385727487f46ebae6a642e3244d1b28  jwks.json

With thanks to GitHub user @craftxbox; closes #17.

ShrimpWink commented:
Hey Awesome code, 

Works great for my vaccine. 
Not that i want to know them or get in trouble but i heard private keys are readily available out there. Is that true ?
I have just heard people now generating QR codes

dlh3 commented:
@ShrimpWink if private keys for trusted passport issuers have leaked, that's concerning and certainly undermines the entire system.  If you know of specific governments/organizations that have had their private key leaked, I would encourage you to notify them, as it means they should stop accepting passports signed with the leaked key and re-issue new ones.

It would also be worth reporting to VCI Commons (https://github.com/the-commons-project/vci-directory) so they can remove those governments/organizations from the trusted issuers list until they re-issue passports signed with a new key.

ShrimpWink commented:
Hey , just a rumour that I heard on the forums

No proof at all. I’m sure it’s just a rumour

Again , great program
On Wed, Oct 20, 2021 at 3:37 PM Dave Hughes ***@***.***>
wrote:

> @ShrimpWink <https://github.com/ShrimpWink> if private keys for trusted
> passport issuers have leaked, that's concerning and certainly undermines
> the entire system. If you know of specific governments/organizations that
> have had their private key leaked, I would encourage you to notify them, as
> it means they should stop accepting passports signed with the leaked key
> and re-issue new ones.
>
> It would also be worth reporting to VCI Commons (
> https://github.com/the-commons-project/vci-directory) so they can remove
> those governments/organizations from the trusted issuers list until they
> re-issue passports signed with a new key.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fproulx/shc-covid19-decoder/pull/6#issuecomment-948054990>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWEV4UZZBWP4HLBL4B7G4UDUH4ZANANCNFSM467LOL7A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

steven676 committed:
Add signing key for Delaware DelVAX immunization records

Delaware appears to use Envision Technology's WebIZ platform [1] for
tracking immunization records.

Key from https://smarthealthcard.iisregistry.net/delaware/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
1f3e96e5a653a0c7ef5c33b3a6f85282ce2605516bb9db9569e8c48e8d830857  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/215.)

[1] https://envisiontechnology.com/products/

steven676 committed:
Add signing key for Connecticut CT WiZ immunzation records

Connecticut appears to use Envision Technology's WebIZ platform [1] for
tracking immunization records.

Key from https://smarthealthcard.iisregistry.net/connecticut/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
4120c091b645e921943b9cf95285797741e60ae003d0df710fbfa7d861e59a7c  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/222.)

[1] https://envisiontechnology.com/products/

steven676 committed:
Add signing key for Nevada WebIZ immunization records

Nevada appears to use Envision Technology's WebIZ platform [1] for
tracking immunization records.

Key from https://smarthealthcard.iisregistry.net/nevada/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
2f503c6a35c59e8426d202a4be136b60f89f00a733b2a976ba548711f964999c  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/226.)

[1] https://envisiontechnology.com/products/

steven676 committed:
Add signing key for New Mexico VaxViewNM immunization records

New Mexico appears to use Envision Technology's WebIZ platform [1] for
tracking immunization records.

Key from https://smarthealthcard.iisregistry.net/newmexico/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
29a5fa3373b91bc5154c8939813da6fae3b4ce7ab795e83b0e0b523b882062d0  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/227.)

[1] https://envisiontechnology.com/products/

steven676 committed:
Add signing key for Oklahoma State Immunization Information System (OSIIS) records

Oklahoma appears to use Envision Technology's WebIZ platform [1] for
tracking immunization records.

Key from https://smarthealthcard.iisregistry.net/oklahoma/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
6e842d2d805b90ec9211594f502f01217c2ddbefc61f9d15056e5043c594e600  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/228.)

[1] https://envisiontechnology.com/products/

steven676 committed:
Add signing key for CNMI Immunization Registry records

The Northern Mariana Islands appear to use Envision Technology's WebIZ
platform [1] for tracking immunization records.

Key from https://smarthealthcard.iisregistry.net/cnmi/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
9e701b6b4703ed83b36541c51e4317d04e50f3b1ad82344feb907ad6e30ea193  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/229.)

[1] https://envisiontechnology.com/products/

steven676 committed:
Add signing key for Guam Immunization Registry (GuWebIZ) records

Guam appears to use Envision Technology's WebIZ platform [1] for
tracking immunization records.

Key from https://smarthealthcard.iisregistry.net/guam/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
225040d69245b29d19ffb762184ce9ac592450042d962252cb810b95827598f4  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/230.)

[1] https://envisiontechnology.com/products/

steven676 committed:
Add signing key for Kentucky Immunization Registry records

Kentucky appears to use Envision Technology's WebIZ platform [1] for
tracking immunization records.

Keys from https://smarthealthcard.iisregistry.net/kentucky/issuer/.well-known/jwks.json:

$ sha256sum jwks.json
7be3f7da7e3b309d238a75c68cad8a4b9c4c835dcdcb56ad5f8624c14dc67e0b  jwks.json

(Found via https://github.com/the-commons-project/vci-directory/pull/231.)

[1] https://envisiontechnology.com/products/

steven676 committed:
Add signing key for Prince Edward Island PEIVaxPass Covid-19 vaccination record

Keys from https://pvcprod.gov.pe.ca/.well-known/jwks.json:

$ sha256sum jwks.json
67ebfe49451fc583c9c92bf9f75d480df5e63a1530a6b88009632c5809732e13  jwks.json

(Found via https://github.com/billylo1/covidpass/commit/901c61dc9808fe09e877e5b22cbb13ae77704031;
with thanks to the grassroots vaccine pass team (@grassroots_team on
Twitter)).

steven676 commented:
More signing keys:

* US states and territories: Delaware, Connecticut, Nevada, New Mexico, Oklahoma, Kentucky, Guam, the Northern Mariana Islands (all users of [Envision's WebIZ system](https://envisiontechnology.com/products/); as with the Docket users, it's unclear how many are currently issuing SHCs, but the submissions to the VCI directory suggest they're coming)
* Canadian provinces: Newfoundland and Labrador (thanks @craftxbox, closes #17), Prince Edward Island
